File: samples/chat-iot-flutter/lib/retrieve_anomalies_tool.dart
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0

import 'dart:convert';

import 'package:dazzle_flutter/dazzle_flutter.dart';

import 'iot_corpus.dart';

/// Tool the LLM calls when the user asks about sensor data.
///
/// Wire-format (OpenAI-compatible):
/// ```
/// retrieve_anomalies(min_from: integer, min_to: integer)
///   → [{start_minute, end_minute, avg_temp_c, max_temp_c, avg_humidity,
///       anomaly_detected, anomaly_type, summary}]
/// ```
///
/// Stays entirely on the snapshot-cache RESP-free path:
///   1. `sset.rangeByScoreDirect` → short IDs (snapshot HIT).
///   2. per-ID `hash.getAllDirect` → payload fields (snapshot HIT).
///
/// Zero RESP in the hot path.
class RetrieveAnomaliesTool extends Tool<TimeRange, List<Map<String, dynamic>>> {
  @override
  String get name => 'retrieve_anomalies';

  @override
  String get description =>
      'Return the sensor windows overlapping [min_from..min_to] from '
      'the on-device Dazzle store. Each row includes averages, anomaly '
      'flag, and a one-line summary. Minutes are 0..2399.';

  @override
  JsonSchema get argsSchema => jsonSchemaObject(
        description: 'Time range (in minutes) to inspect.',
        build: (b) {
          b.property('min_from',
              type: 'integer',
              description: 'Lower-bound minute, inclusive (0..2399).',
              required: true,
              minimum: 0,
              maximum: 2399);
          b.property('min_to',
              type: 'integer',
              description: 'Upper-bound minute, inclusive (0..2399).',
              required: true,
              minimum: 0,
              maximum: 2399);
        },
      );

  @override
  TimeRange argsFromJson(String raw) {
    final m = jsonDecode(raw) as Map<String, dynamic>;
    return TimeRange(
      minFrom: (m['min_from'] as num).toInt(),
      minTo:   (m['min_to']   as num).toInt(),
    );
  }

  @override
  Future<List<Map<String, dynamic>>> invoke(
      TimeRange args, ToolContext ctx) async {
    final client = DazzleServer.shared.client();
    final sset = client.sortedSet(IotCorpus.sortedSetKey);

    // 1) Fast-path range read → short IDs via snapshot cache.
    final ids = sset.rangeByScoreDirect(
      min: args.minFrom.toDouble(),
      max: args.minTo.toDouble(),
    );

    // 2) Hydrate each ID via `hgetAllDirect` — also snapshot HIT.
    final out = <Map<String, dynamic>>[];
    for (final id in ids) {
      final fields = client.hash('${IotCorpus.hashPrefix}$id').getAllDirect();
      if (fields.isEmpty) continue;
      out.add(IoTWindow.fromFields(fields).toJson());
    }
    return out;
  }

  @override
  String returnToJson(List<Map<String, dynamic>> value) => jsonEncode(value);
}

/// Arguments the LLM sends when it calls `retrieve_anomalies`.
class TimeRange {
  const TimeRange({required this.minFrom, required this.minTo});
  final int minFrom;
  final int minTo;
}

File: samples/chat-iot-rn/src/retrieveAnomaliesTool.ts
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
// SPDX-License-Identifier: Apache-2.0

import {
  DazzleServer, JsonSchema, Tool, ToolContext, jsonSchemaObject,
} from 'dazzle-react-native';
import { IotCorpus, IoTWindow } from './iotCorpus';

export interface TimeRange {
  min_from: number;
  min_to: number;
}

/**
 * Tool the LLM calls when the user asks about sensor data.
 *
 * Stays entirely on the snapshot-cache RESP-free path:
 *   1. `rangeByScoreDirect` → short IDs (JSI ~1 µs / sync bridge ~15 µs).
 *   2. per-ID `hgetAllDirect` → payload fields.
 *
 * Zero RESP round-trips in the hot path.
 */
export class RetrieveAnomaliesTool implements Tool<TimeRange, IoTWindow[]> {
  readonly name = 'retrieve_anomalies';
  readonly description =
    'Return the sensor windows overlapping [min_from..min_to] from ' +
    'the on-device Dazzle store. Each row includes averages, anomaly ' +
    'flag, and a one-line summary. Minutes are 0..2399.';

  readonly argsSchema: JsonSchema = jsonSchemaObject(
    { description: 'Time range (in minutes) to inspect.' },
    (b) => {
      b.property('min_from', { type: 'integer', required: true,
        description: 'Lower-bound minute, inclusive (0..2399).',
        minimum: 0, maximum: 2399 });
      b.property('min_to', { type: 'integer', required: true,
        description: 'Upper-bound minute, inclusive (0..2399).',
        minimum: 0, maximum: 2399 });
    },
  );

  argsFromJson(raw: string): TimeRange {
    const o = JSON.parse(raw);
    return { min_from: Number(o.min_from ?? 0), min_to: Number(o.min_to ?? 0) };
  }

  returnToJson(value: IoTWindow[]): string {
    return JSON.stringify(value);
  }

  async invoke(args: TimeRange, _ctx: ToolContext): Promise<IoTWindow[]> {
    const client = DazzleServer.shared.client();
    const sset = client.sortedSet(IotCorpus.sortedSetKey);

    // 1) Fast-path range read → short IDs via snapshot cache.
    const ids = await sset.rangeByScoreDirect(args.min_from, args.min_to);

    // 2) Hydrate each ID via `hgetAllDirect` — also snapshot HIT.
    const out: IoTWindow[] = [];
    for (const id of ids) {
      const fields = await client.hash(`${IotCorpus.hashPrefix}${id}`)
          .getAllDirect();
      const keys = Object.keys(fields);
      if (keys.length === 0) continue;
      out.push(IotCorpus.fromFields(fields as Record<string, string>));
    }
    return out;
  }
}
